### PR TITLE
Add a basic game flow

### PR DESCRIPTION
**Pull request by theowiik**
https://github.com/erikwessman/ciu265/assets/42991003/33b4dbbd-bc76-4996-ba64-5b6dd2f02914

image

### CODE REVIEW

File: weave/Scripts/Main.cs [hunk 1]
```diff
@@ -1,38 +1,82 @@
...
+            {
+                var goal = Instanter.Instantiate<Goal>();
+                CallDeferred("add_child", goal);
+                goal.GlobalPosition = GetRandomCoordinateInView(100);
+                goal.PlayerReachedGoal += OnPlayerReachedGoal;
+                goal.CallDeferred("set", nameof(Player.PlayerId), player.PlayerId);
+            });
+    }
+
+    private Vector2 GetRandomCoordinateInView(float margin)
```
Comment (erikwessman) [inline weave/Scripts/Main.cs:76]: In the future this method (or some other method) should not be able to give the same coordinates that another entity has. Sometimes the player spawns on top of their goal

File: weave/Scripts/Player.cs [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
     public delegate void PlayerShotBulletEventHandler(Node2D bullet, Vector2 globalPosition);
 
-    private const int MovementSpeed = 100;
-    private PackedScene _bombScene = GD.Load<PackedScene>("res://Objects/Bomb.tscn");
+    private const int MovementSpeed = 500;
+
+    [GetNode("Label")]
+    private Label _label;
+
+    public string PlayerId
```
Comment (erikwessman) [inline weave/Scripts/Player.cs:19]: Again, maybe not a good idea to return the label value when getting/setting the player ID

File: weave/Scripts/Utils/EnumerableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+/// </summary>
+public static class EnumerableExtensions
+{
+    /// <summary>
+    ///     Executes the specified action a specified number of times.
+    ///     TODO: Move to GodotSharper.
+    /// </summary>
+    /// <param name="count">The number of times to execute the action.</param>
+    /// <param name="action">The action to execute.</param>
+    public static void TimesDo(this int count, Action action)
```
Comment (erikwessman) [inline weave/Scripts/Utils/EnumerableExtensions.cs:16]: Throw exception for negative values :nerd_face:

**Review by erikwessman — CHANGES_REQUESTED** [review]
Just some small things :smiley:

File: weave/Scripts/Main.cs (hunk 1 above)
Comment (theowiik) [reply weave/Scripts/Main.cs:76]: Yes, this is just a placeholder

File: weave/Scripts/Goal.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    ///     This is the signal that will be emitted when a player reaches the goal.
+    ///     Will only be emitted once.
+    /// </summary>
+    [Signal]
+    public delegate void PlayerReachedGoalEventHandler(Player player);
+
+    [GetNode("Label")]
+    private Label _label;
+
+    public string PlayerId
```
[the remark this answers is not in the document]
Comment (theowiik) [reply weave/Scripts/Goal.cs:21]: Yes true, I was being a bit cheeky 🤓, also assuming that most of this will be altered but will fix

**Review by erikwessman — APPROVED** [review]
:+1:

**Pull request merged**